# Venture Funding with Deep Learning

You work as a risk management associate at Alphabet Soup, a venture capital firm. Alphabet Soup’s business team receives many funding applications from startups every day. This team has asked you to help them create a model that predicts whether applicants will be successful if funded by Alphabet Soup.

The business team has given you a CSV containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. With your knowledge of machine learning and neural networks, you decide to use the features in the provided dataset to create a binary classifier model that will predict whether an applicant will become a successful business. The CSV file contains a variety of information about these businesses, including whether or not they ultimately became successful.

## Instructions:

The steps for this challenge are broken out into the following sections:

* Prepare the data for use on a neural network model.

* Compile and evaluate a binary classification model using a neural network.

* Optimize the neural network model.

### Prepare the Data for Use on a Neural Network Model 

Using your knowledge of Pandas and scikit-learn’s `StandardScaler()`, preprocess the dataset so that you can use it to compile and evaluate the neural network model later.

Open the starter code file, and complete the following data preparation steps:

1. Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.   

2. Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.
 
3. Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

4. Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course. 

5. Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset. 

6. Split the features and target sets into training and testing datasets.

7. Use scikit-learn's `StandardScaler` to scale the features data.

### Compile and Evaluate a Binary Classification Model Using a Neural Network

Use your knowledge of TensorFlow to design a binary classification deep neural network model. This model should use the dataset’s features to predict whether an Alphabet Soup&ndash;funded startup will be successful based on the features in the dataset. Consider the number of inputs before determining the number of layers that your model will contain or the number of neurons on each layer. Then, compile and fit your model. Finally, evaluate your binary classification model to calculate the model’s loss and accuracy. 
 
To do so, complete the following steps:

1. Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.

2. Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.

> **Hint** When fitting the model, start with a small number of epochs, such as 20, 50, or 100.

3. Evaluate the model using the test data to determine the model’s loss and accuracy.

4. Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`. 

### Optimize the Neural Network Model

Using your knowledge of TensorFlow and Keras, optimize your model to improve the model's accuracy. Even if you do not successfully achieve a better accuracy, you'll need to demonstrate at least two attempts to optimize the model. You can include these attempts in your existing notebook. Or, you can make copies of the starter notebook in the same folder, rename them, and code each model optimization in a new notebook. 

> **Note** You will not lose points if your model does not achieve a high accuracy, as long as you make at least two attempts to optimize the model.

To do so, complete the following steps:

1. Define at least three new deep neural network models (the original plus 2 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.

2. After finishing your models, display the accuracy scores achieved by each model, and compare the results.

3. Save each of your models as an HDF5 file.


In [1]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
import warnings
warnings.filterwarnings('ignore')

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.  


In [4]:
# Read the applicants_data.csv file from the Resources folder into a Pandas DataFrame
applicant_data_df = pd.read_csv('./Resources/applicants_data.csv')

# Review the DataFrame
applicant_data_df.head()


EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [5]:
applicant_data_df.tail()

EIN                                               NAME  \
34294  996009318              THE LIONS CLUB OF HONOLULU KAMEHAMEHA   
34295  996010315           INTERNATIONAL ASSOCIATION OF LIONS CLUBS   
34296  996012607                                PTA HAWAII CONGRESS   
34297  996015768  AMERICAN FEDERATION OF GOVERNMENT EMPLOYEES LO...   
34298  996086871                           WATERHOUSE CHARITABLE TR   

      APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
34294   Association       1          0                      N      5000   
34295   Association       1          0                      N      5000   
34296   Association       1          0                      N      5000   
34297   Association       1          0                      N      5000   
34298  Co-operative       1      1M-5M                      N  36500179   

       IS_SUCCESSFUL  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0

In [6]:
# Review the data types associated with the columns
applicant_data_df.dtypes

EIN                        int64
NAME                      object
APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

### Step 2: Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

In [7]:
# Drop the 'EIN' and 'NAME' columns from the DataFrame
applicant_data_df = applicant_data_df.drop(columns = ['EIN', 'NAME'])

# Review the DataFrame
applicant_data_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

### Step 3: Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

In [8]:
# Create a list of categorical variables 
categorical_variables = list(applicant_data_df.dtypes[applicant_data_df.dtypes == "object"].index)

# Display the categorical variables list
categorical_variables

['APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'INCOME_AMT',
 'SPECIAL_CONSIDERATIONS']

In [9]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse = False)

In [10]:
# Encode the categorcal variables using OneHotEncoder
encoded_data = enc.fit_transform(applicant_data_df[categorical_variables])

In [11]:
encoded_data.shape

(34299, 114)

In [12]:
# Create a DataFrame with the encoded variables
encoded_df = pd.DataFrame(
    encoded_data,
    columns = enc.get_feature_names(categorical_variables)
)

# Review the DataFrame
encoded_df.head()

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                   0.0  ...                0.0                     0.0   
1                   0.0  ...                1.0                     0.0   
2                   0.0  ...                0.0                     0.0   
3                   0.0  ...                0.0                     1.0   
4                   0.0  ...                0.0                     0.0   

   INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                       0.0                 0.0               0.0   
1                       0.0                 0.0               0.0   
2                       0.0                 0.0               0.0   
3                       0.0                 0.0               0.0   
4                       1.0                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
0                     0.0              0.0                0.0   
1                     0.0              0.0                0.0   
2                     0.0              0.0                0.0   
3                     0.0              0.0                0.0   
4                     0.0              0.0                0.0   

   SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                       1.0                       0.0  
1                       1.0                       0.0  
2                       1.0                       0.0  
3                       1.0                       0.0  
4                       1.0                       0.0  

[5 rows x 114 columns]

### Step 4: Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course. 

In [13]:
# Add the numerical variables from the original DataFrame to the one-hot encoding DataFrame
numerical_variables_df = applicant_data_df.drop(columns = categorical_variables)

numerical_variables_df.head()


STATUS  ASK_AMT  IS_SUCCESSFUL
0       1     5000              1
1       1   108590              1
2       1     5000              0
3       1     6692              1
4       1   142590              1

In [14]:
encoded_df = pd.concat(
    [encoded_df, numerical_variables_df],
    axis=1
)

# Review the Dataframe
encoded_df.head()

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                   0.0  ...                 0.0               0.0   
1                   0.0  ...                 0.0               0.0   
2                   0.0  ...                 0.0               0.0   
3                   0.0  ...                 0.0               0.0   
4                   0.0  ...                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
0                     0.0              0.0                0.0   
1                     0.0              0.0                0.0   
2                     0.0              0.0                0.0   
3                     0.0              0.0                0.0   
4                     0.0              0.0                0.0   

   SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  STATUS  ASK_AMT  \
0                       1.0                       0.0       1     5000   
1                       1.0                       0.0       1   108590   
2                       1.0                       0.0       1     5000   
3                       1.0                       0.0       1     6692   
4                       1.0                       0.0       1   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1  

[5 rows x 117 columns]

---

### Step 5: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset. 



In [15]:
# Define the target set y using the IS_SUCCESSFUL column
y = encoded_df['IS_SUCCESSFUL']

# Display a sample of y
y[:5]


0    1
1    1
2    0
3    1
4    1
Name: IS_SUCCESSFUL, dtype: int64

In [16]:
# Define features set X by selecting all columns but IS_SUCCESSFUL
X = encoded_df.drop(columns =['IS_SUCCESSFUL'])

# Review the features DataFrame
X[:5]


APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                   0.0  ...                       0.0                 0.0   
1                   0.0  ...                       0.0                 0.0   
2                   0.0  ...                       0.0                 0.0   
3                   0.0  ...                       0.0                 0.0   
4                   0.0  ...                       1.0                 0.0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0               0.0                     0.0              0.0   
1               0.0                     0.0              0.0   
2               0.0                     0.0              0.0   
3               0.0                     0.0              0.0   
4               0.0                     0.0              0.0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  \
0                0.0                       1.0                       0.0   
1                0.0                       1.0                       0.0   
2                0.0                       1.0                       0.0   
3                0.0                       1.0                       0.0   
4                0.0                       1.0                       0.0   

   STATUS  ASK_AMT  
0       1     5000  
1       1   108590  
2       1     5000  
3       1     6692  
4       1   142590  

[5 rows x 116 columns]

### Step 6: Split the features and target sets into training and testing datasets.


In [17]:
# Split the preprocessed data into a training and testing dataset
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


### Step 7: Use scikit-learn's `StandardScaler` to scale the features data.

In [18]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

---

## Compile and Evaluate a Binary Classification Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [19]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

# Review the number of features
number_input_features

116

In [20]:
# Define the number of neurons in the output layer
number_output_neurons = 1

In [21]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = (number_input_features + 1) // 2 

# Review the number hidden nodes in the first layer
hidden_nodes_layer1

58

In [22]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 = (number_input_features + 1) // 2 

# Review the number hidden nodes in the second layer
hidden_nodes_layer2

58

In [23]:
# Create the Sequential model instance
nn = Sequential()

In [24]:
# Add the first hidden layer
nn.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))


In [25]:
# Add the second hidden layer
nn.add(Dense(units=hidden_nodes_layer2, activation="relu"))


In [26]:
# Add the output layer to the model specifying the number of output neurons and activation function
nn.add(Dense(units=1, activation="sigmoid"))


In [27]:
# Display the Sequential model summary
nn.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 58)                6786      
                                                                 
 dense_1 (Dense)             (None, 58)                3422      
                                                                 
 dense_2 (Dense)             (None, 1)                 59        
                                                                 
Total params: 10,267
Trainable params: 10,267
Non-trainable params: 0
_________________________________________________________________


### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [28]:
# Compile the Sequential model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [29]:
# Fit the model using 50 epochs and the training data
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5750 - accuracy: 0.7211
Epoch 2/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5538 - accuracy: 0.7286
Epoch 3/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5502 - accuracy: 0.7321
Epoch 4/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5471 - accuracy: 0.7324
Epoch 5/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5458 - accuracy: 0.7332
Epoch 6/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5457 - accuracy: 0.7344
Epoch 7/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5435 - accuracy: 0.7335
Epoch 8/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5431 - accuracy: 0.7337
Epoch 9/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5427 - accuracy: 0.7351
Epoch 10/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5417 - accuracy: 0.7356

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [30]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5587 - accuracy: 0.7311 - 462ms/epoch - 2ms/step
Loss: 0.5586576461791992, Accuracy: 0.7310787439346313


### Step 4: Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`. 


In [31]:
# Set the model's file path
from pathlib import Path

file_path = Path("./Resources/AlphabetSoup.h5")

# Export your model to a HDF5 file
nn.save(file_path)

---

## Optimize the neural network model


### Step 1: Define at least three new deep neural network models (resulting in the original plus 3 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.


### Alternative Model 1

### Alternative Model 1. Parameters are as follow: 
#### - ***Input Features:*** Keep constant at 116 for this model, to avoid resplitting the train/test data
#### - ***Hidden Layer Neurons:*** Use the same formula/proportion = (input)/2 = 58
#### - ***Hidden Layers:*** Add one more hidden layer - for a total of 3 - and the last one with the same number of nodes = 58
#### - ***Hidden Layer Activation:*** Use same activatation ReLu
#### - ***Epochs :*** Keep them constant at 50.

In [32]:
# Define the the number of inputs (features) to the model
number_input_features_A1 = len(X_train.iloc[0])

# Review the number of features
number_input_features_A1

116

In [33]:
# Define the number of neurons in the output layer
number_output_neurons_A1 = 1

In [34]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A1 = (number_input_features_A1) // 2 

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A1

58

In [35]:
# Define the number of hidden nodes for the SECOND hidden layer
hidden_nodes_layer2_A1 = (number_input_features_A1) // 2 

# Review the number of hidden nodes in the first layer
hidden_nodes_layer2_A1

58

In [36]:
# Define the number of hidden nodes for the THIRD hidden layer
hidden_nodes_layer3_A1 = (number_input_features_A1) // 2 

# Review the number of hidden nodes in the first layer
hidden_nodes_layer3_A1

58

In [37]:
from keras.models import Sequential
from keras.layers import Dense

In [38]:
# Create the Sequential model instance
nn_A1 = Sequential()

In [39]:
# A1 FIRST hidden layer
nn_A1.add(Dense(units=hidden_nodes_layer1_A1, input_dim=number_input_features_A1, activation="relu"))

# A1 SECOND hidden layer
nn_A1.add(Dense(units=hidden_nodes_layer2_A1,activation="relu"))

# A1 THIRD hidden layer
nn_A1.add(Dense(units=hidden_nodes_layer3_A1,activation="relu"))

# Output layer
nn_A1.add(Dense(units=1,activation="sigmoid"))

In [40]:
# Check the structure of the model
nn_A1.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 58)                6786      
                                                                 
 dense_4 (Dense)             (None, 58)                3422      
                                                                 
 dense_5 (Dense)             (None, 58)                3422      
                                                                 
 dense_6 (Dense)             (None, 1)                 59        
                                                                 
Total params: 13,689
Trainable params: 13,689
Non-trainable params: 0
_________________________________________________________________


In [41]:
# Compile the Sequential model
nn_A1.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [42]:
# Fit the model using 50 epochs and the training data
fit_model_A1 = nn_A1.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 2s 1ms/step - loss: 0.5708 - accuracy: 0.7188
Epoch 2/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5512 - accuracy: 0.7298
Epoch 3/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5481 - accuracy: 0.7325
Epoch 4/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5459 - accuracy: 0.7320
Epoch 5/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5445 - accuracy: 0.7333
Epoch 6/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5438 - accuracy: 0.7350
Epoch 7/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5428 - accuracy: 0.7344
Epoch 8/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5420 - accuracy: 0.7357
Epoch 9/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5419 - accuracy: 0.7349
Epoch 10/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5408 - accuracy: 0.7362

#### Alternative Model 2

### Alternative Model 2. Parameters are as follow: 
#### - ***Input Features:*** Keep constant at 116 for this model, to avoid resplitting the train/test data
#### - ***Hidden Layer Neurons:*** Use the same formula/proportion = (input)/2 = 58
#### - ***Hidden Layers:*** Add 2 more hidden layers - for a total of 4 - all with the same number of nodes = 58
#### - ***Hidden Layer Activation:*** Use same activatation ReLu
#### - ***Epochs :*** Increase the Epochs to 100.

In [43]:
# Define the the number of inputs (features) to the model
number_input_features_A2 = len(X_train.iloc[0])

# Review the number of features
number_input_features_A2

116

In [44]:
# Define the number of neurons in the output layer
number_output_neurons_A2 = 1

In [45]:
# Define the number of hidden nodes for the FIRST hidden layer
hidden_nodes_layer1_A2 =  (number_input_features_A2) // 2 

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A2

58

In [46]:
# Define the number of hidden nodes for the SECOND hidden layer
hidden_nodes_layer2_A2 =  hidden_nodes_layer1_A2

# Review the number of hidden nodes in the first layer
hidden_nodes_layer2_A2

58

In [47]:
# Define the number of hidden nodes for the THIRD hidden layer
hidden_nodes_layer3_A2 =  hidden_nodes_layer2_A2

# Review the number of hidden nodes in the first layer
hidden_nodes_layer3_A2

58

In [48]:
# Create the Sequential model instance
nn_A2 = Sequential()

In [49]:
# A2 FIRST hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer1_A2, input_dim=number_input_features_A2, activation="relu"))

# A2 SECOND hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer2_A2, activation="relu"))

# A2 THIRD hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer3_A2,activation="relu"))

# Output layer
nn_A2.add(Dense(units=1,activation="sigmoid"))

In [50]:
# Check the structure of the model
nn_A2.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_7 (Dense)             (None, 58)                6786      
                                                                 
 dense_8 (Dense)             (None, 58)                3422      
                                                                 
 dense_9 (Dense)             (None, 58)                3422      
                                                                 
 dense_10 (Dense)            (None, 1)                 59        
                                                                 
Total params: 13,689
Trainable params: 13,689
Non-trainable params: 0
_________________________________________________________________


In [51]:
# Compile the model
nn_A2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [52]:
# Fit the model
fit_model_A2 = nn_A2.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 1ms/step - loss: 0.5737 - accuracy: 0.7175
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5519 - accuracy: 0.7289
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5483 - accuracy: 0.7319
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5461 - accuracy: 0.7327
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5448 - accuracy: 0.7333
Epoch 6/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5449 - accuracy: 0.7343
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5431 - accuracy: 0.7338
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5425 - accuracy: 0.7352
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5422 - accuracy: 0.7352
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5412 - accura

#### Alternative Model 3

### Alternative Model 3. Parameters are as follow: 
#### - ***Input Features:*** Reduce the number of features/inputs by eliminating those that are unclear e.g. (Application Type and maybe Classification). Re split the train/test data with the simplified data set (encoded_filtered Data Frame) and 
#### - ***Hidden Layer Neurons:*** Use the same formula/proportion = (input)/2 = 58/2 +1 = 30
#### - ***Hidden Layers:*** Add 2 more hidden layers - for a total of 4 - all with the same number of nodes = 30
#### - ***Hidden Layer Activation:*** Use same activatation ReLu
#### - ***Epochs :*** Use the same number of Epochs as the original model = 50.

In [383]:
#Understanding the data in the encoded_df set

col_counts = encoded_df.apply(pd.Series.value_counts)
print(col_counts)

              APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  \
0.000000e+00               33771.0               34272.0   
1.000000e+00                 528.0                  27.0   
5.000000e+03                   NaN                   NaN   
5.001000e+03                   NaN                   NaN   
5.006000e+03                   NaN                   NaN   
...                            ...                   ...   
3.391919e+09                   NaN                   NaN   
4.653012e+09                   NaN                   NaN   
5.591585e+09                   NaN                   NaN   
8.556639e+09                   NaN                   NaN   
8.597806e+09                   NaN                   NaN   

              APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  \
0.000000e+00               34233.0               34296.0   
1.000000e+00                  66.0                   3.0   
5.000000e+03                   NaN                   NaN   
5.001000e+03                   NaN     

In [384]:
# Simplify the dataset by eliminating unncecessary features

# encoded_filtered = encoded_df.filter(regex='^(?!.*CLASSIFICATION|.*AFFILIATION|.*APPLICATION_TYPE).*$', axis=1)

encoded_filtered = encoded_df.filter(regex='^(?!.*APPLICATION_TYPE|.*CLASSIFICATION_).*$', axis=1)

# Verify the simplified data frame

encoded_filtered.head()

AFFILIATION_CompanySponsored  AFFILIATION_Family/Parent  \
0                           0.0                        0.0   
1                           0.0                        0.0   
2                           1.0                        0.0   
3                           1.0                        0.0   
4                           0.0                        0.0   

   AFFILIATION_Independent  AFFILIATION_National  AFFILIATION_Other  \
0                      1.0                   0.0                0.0   
1                      1.0                   0.0                0.0   
2                      0.0                   0.0                0.0   
3                      0.0                   0.0                0.0   
4                      1.0                   0.0                0.0   

   AFFILIATION_Regional  USE_CASE_CommunityServ  USE_CASE_Heathcare  \
0                   0.0                     0.0                 0.0   
1                   0.0                     0.0                 0.0   
2                   0.0                     0.0                 0.0   
3                   0.0                     0.0                 0.0   
4                   0.0                     0.0                 1.0   

   USE_CASE_Other  USE_CASE_Preservation  ...  INCOME_AMT_10M-50M  \
0             0.0                    0.0  ...                 0.0   
1             0.0                    1.0  ...                 0.0   
2             0.0                    0.0  ...                 0.0   
3             0.0                    1.0  ...                 0.0   
4             0.0                    0.0  ...                 0.0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0               0.0                     0.0              0.0   
1               0.0                     0.0              0.0   
2               0.0                     0.0              0.0   
3               0.0                     0.0              0.0   
4               0.0                     0.0              0.0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  \
0                0.0                       1.0                       0.0   
1                0.0                       1.0                       0.0   
2                0.0                       1.0                       0.0   
3                0.0                       1.0                       0.0   
4                0.0                       1.0                       0.0   

   STATUS  ASK_AMT  IS_SUCCESSFUL  
0       1     5000              1  
1       1   108590              1  
2       1     5000              0  
3       1     6692              1  
4       1   142590              1  

[5 rows x 29 columns]

In [385]:
encoded_filtered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34299 entries, 0 to 34298
Data columns (total 29 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   AFFILIATION_CompanySponsored  34299 non-null  float64
 1   AFFILIATION_Family/Parent     34299 non-null  float64
 2   AFFILIATION_Independent       34299 non-null  float64
 3   AFFILIATION_National          34299 non-null  float64
 4   AFFILIATION_Other             34299 non-null  float64
 5   AFFILIATION_Regional          34299 non-null  float64
 6   USE_CASE_CommunityServ        34299 non-null  float64
 7   USE_CASE_Heathcare            34299 non-null  float64
 8   USE_CASE_Other                34299 non-null  float64
 9   USE_CASE_Preservation         34299 non-null  float64
 10  USE_CASE_ProductDev           34299 non-null  float64
 11  ORGANIZATION_Association      34299 non-null  float64
 12  ORGANIZATION_Co-operative     34299 non-null  float64
 13  O

In [386]:
# Prepare the Training and Testing data for the new dataframe
y3 = encoded_filtered['IS_SUCCESSFUL']

#Display y
y3[:5]

0    1
1    1
2    0
3    1
4    1
Name: IS_SUCCESSFUL, dtype: int64

In [387]:
# Define features set X by selecting all columns except the one for y 'IS_SUCCESSFUL'
X3 = encoded_filtered.drop(columns =['IS_SUCCESSFUL'])

# Review the features DataFrame
X3[:5]


AFFILIATION_CompanySponsored  AFFILIATION_Family/Parent  \
0                           0.0                        0.0   
1                           0.0                        0.0   
2                           1.0                        0.0   
3                           1.0                        0.0   
4                           0.0                        0.0   

   AFFILIATION_Independent  AFFILIATION_National  AFFILIATION_Other  \
0                      1.0                   0.0                0.0   
1                      1.0                   0.0                0.0   
2                      0.0                   0.0                0.0   
3                      0.0                   0.0                0.0   
4                      1.0                   0.0                0.0   

   AFFILIATION_Regional  USE_CASE_CommunityServ  USE_CASE_Heathcare  \
0                   0.0                     0.0                 0.0   
1                   0.0                     0.0                 0.0   
2                   0.0                     0.0                 0.0   
3                   0.0                     0.0                 0.0   
4                   0.0                     0.0                 1.0   

   USE_CASE_Other  USE_CASE_Preservation  ...  INCOME_AMT_100000-499999  \
0             0.0                    0.0  ...                       0.0   
1             0.0                    1.0  ...                       0.0   
2             0.0                    0.0  ...                       0.0   
3             0.0                    1.0  ...                       0.0   
4             0.0                    0.0  ...                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  STATUS  ASK_AMT  
0                       0.0       1     5000  
1                       0.0       1   108590  
2                       0.0       1     5000  
3                       0.0       1     6692  
4                       0.0       1   142590  

[5 rows x 28 columns]

In [388]:
# Split the filtered data into a new training and new testing datasets

X3_train, X3_test, y3_train, y3_test = train_test_split(X3, y3, random_state=1)

In [389]:
# Create a new StandardScaler instance
scaler3 = StandardScaler()

# Fit the scaler to the features training dataset
X3_scaler = scaler3.fit(X3_train)

# Fit the scaler to the features training dataset
X3_train_scaled = X3_scaler.transform(X3_train)
X3_test_scaled = X3_scaler.transform(X3_test)

In [390]:
# Define the the number of inputs (features) to the model
number_input_features_A3 = len(X3_train.iloc[0])

# Review the number of features
number_input_features_A3

28

In [391]:
# Define the number of neurons in the output layer
number_output_neurons_A3 = 1

In [392]:
# Define the number of hidden nodes for the FIRST hidden layer
hidden_nodes_layer1_A3 =  (number_input_features_A3) #// 2 

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A3

28

In [393]:
# Define the number of hidden nodes for the SECOND hidden layer
hidden_nodes_layer2_A3 =  hidden_nodes_layer1_A3

# Review the number of hidden nodes in the SECOND layer
hidden_nodes_layer2_A3

28

In [400]:
# Define the number of hidden nodes for the THIRD hidden layer
hidden_nodes_layer3_A3 =  (hidden_nodes_layer2_A3  // 2)

# Review the number of hidden nodes in the THIRD layer
hidden_nodes_layer3_A3

14

In [401]:
# Create the Sequential model instance
nn_A3 = Sequential()

In [402]:
# A3 FIRST hidden layer
nn_A3.add(Dense(units=hidden_nodes_layer1_A3, input_dim=number_input_features_A3, activation="relu"))

# A3 SECOND hidden layer
nn_A3.add(Dense(units=hidden_nodes_layer2_A3, activation="relu"))

# A3 THIRD hidden layer
nn_A3.add(Dense(units=hidden_nodes_layer3_A3, activation="relu"))

# Output layer
nn_A3.add(Dense(units=1,activation="sigmoid"))

In [403]:
# Check the structure of the model
nn_A3.summary()

Model: "sequential_24"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_88 (Dense)            (None, 28)                812       
                                                                 
 dense_89 (Dense)            (None, 28)                812       
                                                                 
 dense_90 (Dense)            (None, 14)                406       
                                                                 
 dense_91 (Dense)            (None, 1)                 15        
                                                                 
Total params: 2,045
Trainable params: 2,045
Non-trainable params: 0
_________________________________________________________________


In [404]:
# Compile the model
nn_A3.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [405]:
# Fit the model
fit_model_A3 = nn_A3.fit(X3_train_scaled, y3_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 2s 1ms/step - loss: 0.6122 - accuracy: 0.6898
Epoch 2/50
804/804 [==============================] - 1s 1ms/step - loss: 0.6022 - accuracy: 0.6968
Epoch 3/50
804/804 [==============================] - 1s 1ms/step - loss: 0.6019 - accuracy: 0.6976
Epoch 4/50
804/804 [==============================] - 1s 1ms/step - loss: 0.6011 - accuracy: 0.6983
Epoch 5/50
804/804 [==============================] - 1s 1ms/step - loss: 0.6006 - accuracy: 0.6988
Epoch 6/50
804/804 [==============================] - 1s 2ms/step - loss: 0.6006 - accuracy: 0.6983
Epoch 7/50
804/804 [==============================] - 1s 2ms/step - loss: 0.6000 - accuracy: 0.6992
Epoch 8/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5999 - accuracy: 0.6996
Epoch 9/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5999 - accuracy: 0.6996
Epoch 10/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5996 - accuracy: 0.7002

### Step 2: After finishing your models, display the accuracy scores achieved by each model, and compare the results.

In [406]:
print("Original Model Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Original Model Results
268/268 - 0s - loss: 0.5587 - accuracy: 0.7311 - 300ms/epoch - 1ms/step
Loss: 0.5586576461791992, Accuracy: 0.7310787439346313


In [407]:
print("Alternative Model 1 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
## Alternative Model 1 evaluation
model_A1_loss, model_A1_accuracy = nn_A1.evaluate(X_test_scaled,y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_A1_loss}, Accuracy: {model_A1_accuracy}")

Alternative Model 1 Results
268/268 - 0s - loss: 0.5646 - accuracy: 0.7312 - 382ms/epoch - 1ms/step
Loss: 0.5646246671676636, Accuracy: 0.731195330619812


In [408]:
print("Alternative Model 2 Results")
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
## Alternative Model 2 evaluation
model_A2_loss, model_A2_accuracy = nn_A2.evaluate(X_test_scaled,y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_A2_loss}, Accuracy: {model_A2_accuracy}")

Alternative Model 2 Results
268/268 - 0s - loss: 0.5777 - accuracy: 0.7300 - 312ms/epoch - 1ms/step
Loss: 0.5777302980422974, Accuracy: 0.7300291657447815


In [409]:
print("Alternative Model 3 Results")
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
## Alternative Model 3 evaluation
model_A3_loss, model_A3_accuracy = nn_A3.evaluate(X3_test_scaled,y3_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_A3_loss}, Accuracy: {model_A3_accuracy}")

Alternative Model 3 Results
268/268 - 0s - loss: 0.5983 - accuracy: 0.7005 - 453ms/epoch - 2ms/step
Loss: 0.5983083248138428, Accuracy: 0.7005248069763184


### Step 3: Save each of your alternative models as an HDF5 file.


In [410]:
# Set the file path for the first alternative model
# Set the model's file path
from pathlib import Path

file_A1_path = Path("./Resources/AlphabetSoup_A1_V2.h5")

# Export your model A1 to a HDF5 file
nn_A1.save(file_A1_path)

In [411]:
# Set the file path for the second alternative model
file_A2_path = Path("./Resources/AlphabetSoup_A2_V2.h5")

# Export your model A2 to a HDF5 file
nn_A2.save(file_A2_path)

In [412]:
# Set the file path for the third alternative model
file_A3_path = Path("./Resources/AlphabetSoup_A3_V2.h5")

# Export your model A3 to a HDF5 file
nn_A2.save(file_A3_path)

In [413]:
from watermark import watermark
print(watermark())
print(watermark(iversions=True, globals_=globals(), packages=""))

Last updated: 2023-06-19T14:58:05.691582-07:00

Python implementation: CPython
Python version       : 3.7.13
IPython version      : 7.31.1

Compiler    : MSC v.1916 64 bit (AMD64)
OS          : Windows
Release     : 10
Machine     : AMD64
Processor   : Intel64 Family 6 Model 154 Stepping 4, GenuineIntel
CPU cores   : 12
Architecture: 64bit

tensorflow: 2.11.0
pandas    : 1.3.5

